ls1intum/Artemis
# Feature Toggles: Programming Exercises

feature


### Checklist

### Motivation and Context


We want to have toggles for specific features. In this case, the requirement is to allow admins to disable programming exercises related features that depend on the build and VCS servers. 

### Description

We added new feature toggle services in the client and server, which are generalized, so that  they can be easily extended in case we want to have this functionality for other features in the future. 
There also is a new annotation in Java, which can be set for any classes or methods in REST controllers. Incoming requests will then be blocked (forbidden) if the specified feature is not activated.

### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Select "Feature Toggles" and deactivate a feature
4. The related buttons should now be deactivated for all users

### Screenshots


![feature_toggles](https://user-images.githubusercontent.com/27979674/68869505-9256b200-06f9-11ea-9420-bf1c45ab5d6b.gif)
